particle-iot/device-os
# [gen3] bootloader updates through MBR

### Problem

It's currently possible to brick a device if it's reset or loses power during the bootloader update.

### Solution

This PR resolves this problem and makes bootloader updates safe for Gen 3 devices by using MBR (tiny bootloader bundled with Nordic SoftDevice) to flash the bootloader for us.

Due to the fact that MBR can only reliably access internal flash and the OTA section is in external QSPI flash, we need a temporary location of up to 48K bytes in internal flash for storage. We don't have such an area, so we are borrowing a part of the application area, and the update is performed in the following steps:

1. Bootloader is received into the OTA section.
2. It's added to the module slots in DCT with a patched destination address of the user application (0xd4000) and the device resets
3. Bootloader cycles through the module slots, if needed places the bootloader into the last slot (because update through MBR will trigger a reset)
4. Bootloader copies a part of the application which will be overwritten with the bootloader temporarily into a reserved space in the external flash
5. Bootloader schedules a copy of this partial application back into its place by adding it into the module slots,
6. Bootloader then copies the pending bootloader binary in place of the application and asks MBR to perform an update
7. The device resets, goes back into the bootloader (now already updated)
8. The bootloader normally looks through the module slots, finds a copy pending from step 5 and restores the application
9. We continue to boot into the system as usual

### Steps to Test

1. Build system-part1 out of this branch and flash it onto the device
2. Build bootloader out of this branch (it should have module version 1001), flash OTA or using YModem and make sure that the device takes the update (e.g. particle serial inspect)
3. Modify `system_module_version.mk` and set `BOOTLOADER_VERSION` to 1002.
4. Rebuild the bootloader and flash it OTA or using YModem onto the device.
5. Validate that the bootloader has been updated (e.g. particle serial inspect)

### Example App

N/A

### References

- [CH50688]

---

### Completeness
